File: gitdot-api/src/endpoint/repository/create_repository_commit_filter.rs
use serde::{Deserialize, Serialize};

use crate::{endpoint::Endpoint, resource::RepositoryCommitFilterResource};

pub struct CreateRepositoryCommitFilter;

impl Endpoint for CreateRepositoryCommitFilter {
    const PATH: &'static str = "/repository/{owner}/{repo}/commit_filters";
    const METHOD: http::Method = http::Method::POST;

    type Request = CreateRepositoryCommitFilterRequest;
    type Response = CreateRepositoryCommitFilterResponse;
}

#[derive(ApiRequest, Debug, Serialize, Deserialize)]
pub struct CreateRepositoryCommitFilterRequest {
    pub name: String,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub authors: Option<Vec<String>>,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tags: Option<Vec<String>>,

    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub paths: Option<Vec<String>>,
}

pub type CreateRepositoryCommitFilterResponse = RepositoryCommitFilterResource;

File: gitdot-web/packages/gitdot-api/src/endpoint/repository/create-repository-commit-filter.ts
import { z } from "zod";
import { RepositoryCommitFilterResource } from "../../resource";
import type { Endpoint } from "../endpoint";

export const CreateRepositoryCommitFilterRequest = z.object({
  name: z.string(),
  authors: z.array(z.string()).optional(),
  tags: z.array(z.string()).optional(),
  paths: z.array(z.string()).optional(),
});
export type CreateRepositoryCommitFilterRequest = z.infer<
  typeof CreateRepositoryCommitFilterRequest
>;

export const CreateRepositoryCommitFilterResponse =
  RepositoryCommitFilterResource;
export type CreateRepositoryCommitFilterResponse = z.infer<
  typeof CreateRepositoryCommitFilterResponse
>;

export const CreateRepositoryCommitFilter = {
  path: "/repository/{owner}/{repo}/commit-filters",
  method: "POST",
  request: CreateRepositoryCommitFilterRequest,
  response: CreateRepositoryCommitFilterResponse,
} as const satisfies Endpoint;
export type CreateRepositoryCommitFilter = typeof CreateRepositoryCommitFilter;
